lib/apiservers/engine/backends/container.go

```diff
@@ -828,11 +834,17 @@ func (c *Container) ContainerRm(name string, config *types.ContainerRmConfig) er
 		}
 	}
 
-	if err := c.containerProxy.Remove(vc, config); err != nil {
-		return err
+	// Retry remove operation if container is not in running state.  If in running state, we only try
+	// once to prevent retries from degrading performance.
```

Comment:
My preference would be to perform the operation correctly, even if it takes a little longer, over returning an error. I think that's the essence of the `rm -f` option.

Reply:
It does perform correctly.  It simply makes allowance for a container that we think is running.  In the force removal scenario, we always retry, assuming we were able to stop the container.

Reply:
Retry should be for errors that indicate a transient or race condition, that's why we provide a function that determines whether a retry should be attempted. There's a good default set in the tasks pacakge. Certainly any `CannotDeleteBecauseVMIsPoweredOn` error should not trigger a retry for ContainerRm.

I can see why you'd suppress retry for powered on VMs if not retrying based on error return, but adding additional state based decisions when the state is inherently stale, seems fragile. If we cannot use retry to filter which conditions to retry on (because the portlayer doesn't return the appropriate level of granularity or similar) then we may have to use the approach here for now.

I think we should revisit how we're using handles and the amount of special handling sometime soon - I'd expected this function to look a lot more like the following before looking at this issue (obviously pseudo code with marginal error handling) and without the various special case handling. Similar issues have impacted Exec as well:
```go
var stopErr error
if force {
  stopErr = Stop(id)
  if stopErr == NotFound {
     return errors.New("pretty not found message")
  }
  if stopErr != AlreadyStopped {
     log.Error(stopErr )
     // we're forcing matters so just try the removal anyway
  }
}

err = exec.Remove(id)
if err == NotFound {
   return errors.New("pretty not found message")
} else if err == ContainerIsRunning {
  if stopErr != nil {
    return stopErr 
  }
   return errors.New("must use -f")
}
```